// @progress/kendo-angular-tooltip/esm2020/popover.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { PopupService } from "@progress/kendo-angular-popup";
import { ResizeBatchService } from "@progress/kendo-angular-common";
import { KENDO_POPOVER } from "./directives";
import * as i0 from "@angular/core";
import * as i1 from "./popover/popover.component";
import * as i2 from "./popover/template-directives/actions-template.directive";
import * as i3 from "./popover/template-directives/body-template.directive";
import * as i4 from "./popover/template-directives/title-template.directive";
import * as i5 from "./popover/anchor.directive";
import * as i6 from "./popover/container.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Popover component.
 *
 * @example
 *
 * ```ts-no-run
 * import { PopoverModule } from '@progress/kendo-angular-tooltip';
 *
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 *
 * import { AppComponent } from './app.component';
 *
 * _@NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, PopoverModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class PopoverModule {
}
PopoverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: PopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PopoverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: PopoverModule, imports: [i1.PopoverComponent, i2.PopoverActionsTemplateDirective, i3.PopoverBodyTemplateDirective, i4.PopoverTitleTemplateDirective, i5.PopoverAnchorDirective, i6.PopoverContainerDirective], exports: [i1.PopoverComponent, i2.PopoverActionsTemplateDirective, i3.PopoverBodyTemplateDirective, i4.PopoverTitleTemplateDirective, i5.PopoverAnchorDirective, i6.PopoverContainerDirective] });
PopoverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: PopoverModule, providers: [PopupService, ResizeBatchService], imports: [i1.PopoverComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: PopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_POPOVER],
                    imports: [...KENDO_POPOVER],
                    providers: [PopupService, ResizeBatchService]
                }]
        }] });
