// @progress/kendo-angular-tooltip/esm2020/tooltips.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import { KENDO_TOOLTIPS } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./tooltip/tooltip.directive";
import * as i2 from "./tooltip/tooltip.content.component";
import * as i3 from "./localization/localized-messages.directive";
import * as i4 from "./popover/popover.component";
import * as i5 from "./popover/template-directives/actions-template.directive";
import * as i6 from "./popover/template-directives/body-template.directive";
import * as i7 from "./popover/template-directives/title-template.directive";
import * as i8 from "./popover/anchor.directive";
import * as i9 from "./popover/container.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Tooltips components.
 *
 * @example
 *
 * ```ts-no-run
 * import { TooltipsModule } from '@progress/kendo-angular-tooltip';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * _@NgModule{{
 *    declarations: [AppComponent], // declare app component
 *    imports:      [BrowserModule, TooltipsModule], // import TooltipsModule module
 *    bootstrap:    [AppComponent]
 * }}
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 * ```
 */
export class TooltipsModule {
}
TooltipsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TooltipsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TooltipsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: TooltipsModule, imports: [i1.TooltipDirective, i2.TooltipContentComponent, i3.LocalizedMessagesDirective, i4.PopoverComponent, i5.PopoverActionsTemplateDirective, i6.PopoverBodyTemplateDirective, i7.PopoverTitleTemplateDirective, i8.PopoverAnchorDirective, i9.PopoverContainerDirective], exports: [i1.TooltipDirective, i2.TooltipContentComponent, i3.LocalizedMessagesDirective, i4.PopoverComponent, i5.PopoverActionsTemplateDirective, i6.PopoverBodyTemplateDirective, i7.PopoverTitleTemplateDirective, i8.PopoverAnchorDirective, i9.PopoverContainerDirective] });
TooltipsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TooltipsModule, providers: [PopupService, ResizeBatchService, IconsService], imports: [i2.TooltipContentComponent, i4.PopoverComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TooltipsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_TOOLTIPS],
                    exports: [...KENDO_TOOLTIPS],
                    providers: [PopupService, ResizeBatchService, IconsService]
                }]
        }] });
