// @progress/kendo-angular-scheduler/esm2020/views/day-time/day-time.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { TimeSlotDirective, DaySlotDirective } from './event-slot.directive';
import { DayTimeViewItemComponent } from './day-time-view-item.component';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
const DIRECTIVES = [TimeSlotDirective, DaySlotDirective, DayTimeViewItemComponent];
/**
 * @hidden
 *
 * As of package v17, the `DayTimeModule` is deprecated and can
 * be removed in a future major version. We recommend using `SchedulerModule`.
 */
export class DayTimeModule {
}
DayTimeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DayTimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DayTimeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: DayTimeModule, imports: [TimeSlotDirective, DaySlotDirective, DayTimeViewItemComponent], exports: [TimeSlotDirective, DaySlotDirective, DayTimeViewItemComponent] });
DayTimeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DayTimeModule, providers: [
        IconsService,
        PopupService,
        ResizeBatchService
    ], imports: [DayTimeViewItemComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DayTimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [DIRECTIVES],
                    exports: [DIRECTIVES],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
