// @progress/kendo-angular-scheduler/esm2020/views/month/month-view.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { MonthViewComponent } from './month-view.component';
import { MultiWeekViewComponent } from './multi-week-view.component';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
/**
 * @hidden
 *
 * As of package v17, the `MonthViewModule` is deprecated and can
 * be removed in a future major version. We recommend using `SchedulerModule`.
 */
export class MonthViewModule {
}
MonthViewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MonthViewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MonthViewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: MonthViewModule, imports: [MonthViewComponent,
        MultiWeekViewComponent], exports: [MonthViewComponent,
        MultiWeekViewComponent] });
MonthViewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MonthViewModule, providers: [
        IconsService,
        PopupService,
        ResizeBatchService
    ], imports: [MonthViewComponent,
        MultiWeekViewComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MonthViewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MonthViewComponent,
                        MultiWeekViewComponent
                    ],
                    exports: [
                        MonthViewComponent,
                        MultiWeekViewComponent
                    ],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
