// @progress/kendo-angular-dateinputs/esm2020/common/adaptive.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_BUTTON } from '@progress/kendo-angular-buttons';
import { KENDO_NAVIGATION } from '@progress/kendo-angular-navigation';
import { KENDO_RESIZESENSOR, ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-common";
import * as i2 from "@progress/kendo-angular-buttons";
import * as i3 from "@progress/kendo-angular-navigation";
const KENDO_ADAPTIVE = [
    ...KENDO_RESIZESENSOR,
    ...KENDO_BUTTON,
    ...KENDO_NAVIGATION
];
//IMPORTANT: This module is not needed and can be removed in the following major release.
//The reason to keep it is that the module used to be public prior the migration to standalone.
/**
 * @hidden
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the modules specific to the Adaptive Mode.
 */
export class AdaptiveModule {
}
AdaptiveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: AdaptiveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AdaptiveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: AdaptiveModule, imports: [i1.ResizeSensorComponent, i2.ButtonComponent, i3.ActionSheetComponent, i3.ActionSheetHeaderTemplateDirective, i3.ActionSheetItemTemplateDirective, i3.ActionSheetContentTemplateDirective, i3.ActionSheetFooterTemplateDirective, i3.ActionSheetTemplateDirective, i3.AppBarComponent, i3.AppBarSectionComponent, i3.AppBarSpacerComponent, i3.BottomNavigationComponent, i3.BottomNavigationItemTemplateDirective, i3.BreadCrumbComponent, i3.BreadCrumbItemTemplateDirective], exports: [i1.ResizeSensorComponent, i2.ButtonComponent, i3.ActionSheetComponent, i3.ActionSheetHeaderTemplateDirective, i3.ActionSheetItemTemplateDirective, i3.ActionSheetContentTemplateDirective, i3.ActionSheetFooterTemplateDirective, i3.ActionSheetTemplateDirective, i3.AppBarComponent, i3.AppBarSectionComponent, i3.AppBarSpacerComponent, i3.BottomNavigationComponent, i3.BottomNavigationItemTemplateDirective, i3.BreadCrumbComponent, i3.BreadCrumbItemTemplateDirective] });
AdaptiveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: AdaptiveModule, providers: [
        ResizeBatchService,
        IconsService
    ], imports: [i1.ResizeSensorComponent, i2.ButtonComponent, i3.ActionSheetComponent, i3.AppBarComponent, i3.AppBarSectionComponent, i3.AppBarSpacerComponent, i3.BottomNavigationComponent, i3.BreadCrumbComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: AdaptiveModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_ADAPTIVE],
                    exports: [...KENDO_ADAPTIVE],
                    providers: [
                        ResizeBatchService,
                        IconsService
                    ]
                }]
        }] });
