// @progress/kendo-angular-scheduler/esm2020/views/timeline/timeline-view.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { TimelineViewComponent } from './timeline-view.component';
import { TimelineWeekViewComponent } from './timeline-week-view.component';
import { TimelineMonthViewComponent } from './timeline-month-view.component';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
const PUBLIC_DIRECTIVES = [
    TimelineViewComponent,
    TimelineWeekViewComponent,
    TimelineMonthViewComponent
];
/**
 * @hidden
 *
 * As of package v17, the `TimelineViewModule` is deprecated and can
 * be removed in a future major version. We recommend using `SchedulerModule`.
 */
export class TimelineViewModule {
}
TimelineViewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TimelineViewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TimelineViewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: TimelineViewModule, imports: [TimelineViewComponent,
        TimelineWeekViewComponent,
        TimelineMonthViewComponent], exports: [TimelineViewComponent,
        TimelineWeekViewComponent,
        TimelineMonthViewComponent] });
TimelineViewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TimelineViewModule, providers: [
        IconsService,
        PopupService,
        ResizeBatchService
    ], imports: [PUBLIC_DIRECTIVES] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: TimelineViewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PUBLIC_DIRECTIVES],
                    exports: [PUBLIC_DIRECTIVES],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
