// @progress/kendo-angular-common/esm2020/resize-sensor/resize-sensor.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { ResizeSensorComponent } from './resize-sensor.component';
import { ResizeBatchService } from './resize-batch.service';
import * as i0 from "@angular/core";
/**
 * @hidden
 * IMPORTANT: NgModule export kept for backwards compatibility
 */
export class ResizeSensorModule {
}
ResizeSensorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ResizeSensorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ResizeSensorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: ResizeSensorModule, imports: [ResizeSensorComponent], exports: [ResizeSensorComponent] });
ResizeSensorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ResizeSensorModule, providers: [ResizeBatchService], imports: [ResizeSensorComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ResizeSensorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [ResizeSensorComponent],
                    exports: [ResizeSensorComponent],
                    providers: [ResizeBatchService]
                }]
        }] });
