// @progress/kendo-angular-dateinputs/esm2020/dateinputs.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DATEINPUTS } from './directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { CenturyViewService } from './calendar/services/century-view.service';
import { DecadeViewService } from './calendar/services/decade-view.service';
import { CalendarDOMService } from './calendar/services/dom.service';
import { MonthViewService } from './calendar/services/month-view.service';
import { NavigationService } from './calendar/services/navigation.service';
import { YearViewService } from './calendar/services/year-view.service';
import { DayPeriodService } from './timepicker/services/dayperiod.service';
import { TimePickerDOMService } from './timepicker/services/dom.service';
import { HoursService } from './timepicker/services/hours.service';
import { MillisecondsService } from './timepicker/services/milliseconds.service';
import { MinutesService } from './timepicker/services/minutes.service';
import { SecondsService } from './timepicker/services/seconds.service';
import * as i0 from "@angular/core";
import * as i1 from "./calendar/localization/calendar-custom-messages.component";
import * as i2 from "./calendar/calendar.component";
import * as i3 from "./calendar/templates/cell-template.directive";
import * as i4 from "./calendar/templates/century-cell-template.directive";
import * as i5 from "./calendar/templates/decade-cell-template.directive";
import * as i6 from "./calendar/templates/footer-template.directive";
import * as i7 from "./calendar/templates/header-template.directive";
import * as i8 from "./calendar/templates/header-title-template.directive";
import * as i9 from "./calendar/templates/month-cell-template.directive";
import * as i10 from "./calendar/templates/navigation-item-template.directive";
import * as i11 from "./calendar/templates/weeknumber-cell-template.directive";
import * as i12 from "./calendar/templates/year-cell-template.directive";
import * as i13 from "./calendar/localization/multiview-calendar-custom-messages.component";
import * as i14 from "./calendar/multiview-calendar.component";
import * as i15 from "./dateinput/localization/dateinput-custom-messages.component";
import * as i16 from "./dateinput/dateinput.component";
import * as i17 from "./datepicker/localization/datepicker-custom-messages.component";
import * as i18 from "./datepicker/datepicker.component";
import * as i19 from "./daterange/localization/daterange-popup-custom-messages.component";
import * as i20 from "./daterange/date-range-end-input.directive";
import * as i21 from "./daterange/date-range-popup-template.directive";
import * as i22 from "./daterange/date-range.component";
import * as i23 from "./daterange/date-range-popup.component";
import * as i24 from "./daterange/date-range-selection.directive";
import * as i25 from "./daterange/date-range-start-input.directive";
import * as i26 from "./datetimepicker/localization/datetimepicker-custom-messages.component";
import * as i27 from "./datetimepicker/datetimepicker.component";
import * as i28 from "./timepicker/localization/timepicker-custom-messages.component";
import * as i29 from "./timepicker/timepicker.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Date Inputs components.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Date Inputs module
 * import { DateInputsModule } from '@progress/kendo-angular-dateinputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare the app component
 *     imports:      [BrowserModule, DateInputsModule], // import the Date Inputs module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class DateInputsModule {
}
DateInputsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DateInputsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateInputsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: DateInputsModule, imports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.MultiViewCalendarCustomMessagesComponent, i14.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i17.DatePickerCustomMessagesComponent, i18.DatePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i19.DateRangePopupCustomMessagesComponent, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i20.DateRangeEndInputDirective, i21.DateRangePopupTemplateDirective, i22.DateRangeComponent, i23.DateRangePopupComponent, i24.DateRangeSelectionDirective, i25.DateRangeStartInputDirective, i26.DateTimePickerCustomMessagesComponent, i27.DateTimePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i28.TimePickerCustomMessagesComponent, i29.TimePickerComponent], exports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i13.MultiViewCalendarCustomMessagesComponent, i14.MultiViewCalendarComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i17.DatePickerCustomMessagesComponent, i18.DatePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i10.NavigationItemTemplateDirective, i19.DateRangePopupCustomMessagesComponent, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i20.DateRangeEndInputDirective, i21.DateRangePopupTemplateDirective, i22.DateRangeComponent, i23.DateRangePopupComponent, i24.DateRangeSelectionDirective, i25.DateRangeStartInputDirective, i26.DateTimePickerCustomMessagesComponent, i27.DateTimePickerComponent, i3.CellTemplateDirective, i4.CenturyCellTemplateDirective, i5.DecadeCellTemplateDirective, i6.FooterTemplateDirective, i7.HeaderTemplateDirective, i8.HeaderTitleTemplateDirective, i9.MonthCellTemplateDirective, i10.NavigationItemTemplateDirective, i11.WeekNumberCellTemplateDirective, i12.YearCellTemplateDirective, i28.TimePickerCustomMessagesComponent, i29.TimePickerComponent] });
DateInputsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DateInputsModule, providers: [
        // CalendarModule providers
        IconsService,
        ResizeBatchService,
        CalendarDOMService,
        CenturyViewService,
        DecadeViewService,
        MonthViewService,
        YearViewService,
        PopupService,
        NavigationService,
        // TimePickerModule providers
        TimePickerDOMService,
        HoursService,
        MinutesService,
        SecondsService,
        MillisecondsService,
        DayPeriodService
    ], imports: [i1.CalendarCustomMessagesComponent, i2.CalendarComponent, i13.MultiViewCalendarCustomMessagesComponent, i14.MultiViewCalendarComponent, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i17.DatePickerCustomMessagesComponent, i18.DatePickerComponent, i19.DateRangePopupCustomMessagesComponent, i15.DateInputCustomMessagesComponent, i16.DateInputComponent, i22.DateRangeComponent, i23.DateRangePopupComponent, i26.DateTimePickerCustomMessagesComponent, i27.DateTimePickerComponent, i28.TimePickerCustomMessagesComponent, i29.TimePickerComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: DateInputsModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DATEINPUTS],
                    imports: [...KENDO_DATEINPUTS],
                    providers: [
                        // CalendarModule providers
                        IconsService,
                        ResizeBatchService,
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        PopupService,
                        NavigationService,
                        // TimePickerModule providers
                        TimePickerDOMService,
                        HoursService,
                        MinutesService,
                        SecondsService,
                        MillisecondsService,
                        DayPeriodService
                    ]
                }]
        }] });
