// @progress/kendo-angular-inputs/esm2020/rating/directives/rating-item.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Renders the rating item content. To define an item template, nest an `<ng-template>` tag
 * with the `kendoRatingItemTemplate` directive inside the `<kendo-rating>` tag
 * [see example](slug:templates_rating).
 * The index of the current item is available as an implicit context using the `let-index="index"` syntax.
 */
export class RatingItemTemplateDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
RatingItemTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: RatingItemTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
RatingItemTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.10", type: RatingItemTemplateDirective, isStandalone: true, selector: "[kendoRatingItemTemplate]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: RatingItemTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoRatingItemTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
