// @progress/kendo-angular-scheduler/esm2020/views/common/views-shared.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { ViewFooterComponent } from './view-footer.component';
import { WorkHoursFooterDirective } from './work-hours-footer.directive';
import { RepeatPipe } from './repeat.pipe';
import { ResourceIteratorPipe } from './resource-iterator.pipe';
import { HintContainerComponent } from './hint-container.component';
import { ResizeHintComponent } from './resize-hint.component';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
const DECLARATIONS = [
    ViewFooterComponent,
    WorkHoursFooterDirective,
    RepeatPipe,
    ResourceIteratorPipe,
    HintContainerComponent,
    ResizeHintComponent
];
/**
 * @hidden
 *
 * As of package v17, the `ViewsSharedModule` is deprecated and can
 * be removed in a future major version. We recommend using `SchedulerModule`.
 */
export class ViewsSharedModule {
}
ViewsSharedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ViewsSharedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ViewsSharedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: ViewsSharedModule, imports: [ViewFooterComponent,
        WorkHoursFooterDirective,
        RepeatPipe,
        ResourceIteratorPipe,
        HintContainerComponent,
        ResizeHintComponent], exports: [ViewFooterComponent,
        WorkHoursFooterDirective,
        RepeatPipe,
        ResourceIteratorPipe,
        HintContainerComponent,
        ResizeHintComponent] });
ViewsSharedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ViewsSharedModule, providers: [
        IconsService,
        PopupService,
        ResizeBatchService
    ], imports: [ViewFooterComponent,
        HintContainerComponent,
        ResizeHintComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: ViewsSharedModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [DECLARATIONS],
                    exports: [DECLARATIONS],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
