// @progress/kendo-angular-scheduler/esm2020/views/common/work-hours-footer.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Input } from '@angular/core';
import { ViewFooterComponent } from './view-footer.component';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { clockIcon } from '@progress/kendo-svg-icons';
import * as i0 from "@angular/core";
import * as i1 from "./view-footer.component";
import * as i2 from "@progress/kendo-angular-l10n";
/**
 * @hidden
 */
export class WorkHoursFooterDirective {
    constructor(footer, localization) {
        this.footer = footer;
        this.localization = localization;
        this.showWorkHours = false;
        this.footerItems = [{ cssClass: 'k-scheduler-fullday', fontIcon: 'clock', svgIcon: clockIcon, text: '' }];
    }
    ngOnInit() {
        this.toggleWorkHours();
        this.footer.items = this.footerItems;
    }
    ngOnChanges() {
        this.toggleWorkHours();
    }
    toggleWorkHours() {
        this.footerItems[0].text = this.showWorkHours ? this.localization.get('showFullDay') : this.localization.get('showWorkDay');
    }
}
WorkHoursFooterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: WorkHoursFooterDirective, deps: [{ token: i1.ViewFooterComponent }, { token: i2.LocalizationService }], target: i0.ɵɵFactoryTarget.Directive });
WorkHoursFooterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.2.10", type: WorkHoursFooterDirective, isStandalone: true, selector: "[kendoWorkHoursFooter]", inputs: { showWorkHours: "showWorkHours" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: WorkHoursFooterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoWorkHoursFooter]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ViewFooterComponent }, { type: i2.LocalizationService }]; }, propDecorators: { showWorkHours: [{
                type: Input
            }] } });
