// @progress/kendo-angular-scheduler/esm2020/views/constants.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/** @hidden */
export const MIDNIGHT_INVARIANT = new Date(1980, 0, 1);
/** @hidden */
export const INVARIANT_END = new Date(1980, 0, 2);
/** @hidden */
export const MS_PER_SECOND = 1000;
/** @hidden */
export const MS_PER_MINUTE = 60 * MS_PER_SECOND;
/** @hidden */
export const MS_PER_HOUR = 60 * MS_PER_MINUTE;
/** @hidden */
export const MS_PER_DAY = 24 * MS_PER_HOUR;
//probably should calculate this
/** @hidden */
export const BORDER_WIDTH = 1;
/** @hidden */
export const DEFAULT_EVENT_HEIGHT = 25;
/** @hidden */
export const ONGOING_EVENT_CSS_CLASS = 'k-event-ongoing';
/** @hidden */
export const DAYS_IN_WEEK_COUNT = 7;
/** @hidden */
export const WEEKS_COUNT = 6;
