// @progress/kendo-angular-scheduler/esm2020/views/multi-day/multi-day-view.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { DayViewComponent } from './day-view.component';
import { MultiDayViewComponent } from './multi-day-view.component';
import { WeekViewComponent } from './week-view.component';
import { WorkWeekViewComponent } from './work-week-view.component';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
const PUBLIC_DIRECTIVES = [
    DayViewComponent,
    MultiDayViewComponent,
    WeekViewComponent,
    WorkWeekViewComponent
];
/**
 * @hidden
 *
 * As of package v17, the `MultiDayViewModule` is deprecated and can
 * be removed in a future major version. We recommend using `SchedulerModule`.
 */
export class MultiDayViewModule {
}
MultiDayViewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MultiDayViewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MultiDayViewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: MultiDayViewModule, imports: [DayViewComponent,
        MultiDayViewComponent,
        WeekViewComponent,
        WorkWeekViewComponent], exports: [DayViewComponent,
        MultiDayViewComponent,
        WeekViewComponent,
        WorkWeekViewComponent] });
MultiDayViewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MultiDayViewModule, providers: [
        IconsService,
        PopupService,
        ResizeBatchService
    ], imports: [PUBLIC_DIRECTIVES] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: MultiDayViewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PUBLIC_DIRECTIVES],
                    exports: [PUBLIC_DIRECTIVES],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService
                    ]
                }]
        }] });
