// @progress/kendo-angular-scheduler/esm2020/scheduler.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_SCHEDULER } from './directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { DialogContainerService, DialogService, WindowContainerService, WindowService, } from '@progress/kendo-angular-dialog';
import { CalendarDOMService, CenturyViewService, DayPeriodService, DecadeViewService, HoursService, MillisecondsService, MinutesService, MonthViewService, NavigationService, SecondsService, TimePickerDOMService, YearViewService, } from '@progress/kendo-angular-dateinputs';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { ViewContextService } from './views/view-context.service';
import { ViewStateService } from './views/view-state.service';
import { EditService } from './editing/edit.service';
import { ToolbarService } from './toolbar/toolbar.service';
import * as i0 from "@angular/core";
import * as i1 from "./scheduler.component";
import * as i2 from "./localization/custom-messages.component";
import * as i3 from "./views/scheduler-view.directive";
import * as i4 from "./data-binding.directive";
import * as i5 from "./views/common/slot-selectable.directive";
import * as i6 from "./pdf/pdf.component";
import * as i7 from "./pdf/pdf-command.directive";
import * as i8 from "./navigation/shortcuts.directive";
import * as i9 from "./views/agenda/agenda-view.component";
import * as i10 from "./views/multi-day/day-view.component";
import * as i11 from "./views/month/month-view.component";
import * as i12 from "./views/multi-day/multi-day-view.component";
import * as i13 from "./views/month/multi-week-view.component";
import * as i14 from "./views/timeline/timeline-month-view.component";
import * as i15 from "./views/timeline/timeline-view.component";
import * as i16 from "./views/timeline/timeline-week-view.component";
import * as i17 from "./views/multi-day/week-view.component";
import * as i18 from "./views/multi-day/work-week-view.component";
import * as i19 from "./views/year/year-view.component";
import * as i20 from "./editing-directives/reactive-editing.directive";
import * as i21 from "./editing/recurrence/recurrence-editor.component";
import * as i22 from "./editing/timezone-editor.component";
import * as i23 from "./toolbar/navigation.component";
import * as i24 from "./toolbar/toolbar-template.directive";
import * as i25 from "./toolbar/view-selector.component";
import * as i26 from "./views/templates/agenda-date-template.directive";
import * as i27 from "./views/templates/agenda-time-template.directive";
import * as i28 from "./views/templates/all-day-event-template.directive";
import * as i29 from "./views/templates/all-day-slot-template.directive";
import * as i30 from "./views/templates/date-header-template.directive";
import * as i31 from "./views/templates/event-template.directive";
import * as i32 from "./editing/edit-dialog-template.directive";
import * as i33 from "./views/templates/group-header-template.directive";
import * as i34 from "./views/templates/major-time-header-template.directive";
import * as i35 from "./views/templates/minor-time-header-template.directive";
import * as i36 from "./views/templates/month-day-slot-template.directive";
import * as i37 from "./views/templates/multi-week-day-slot-template.directive";
import * as i38 from "./views/templates/time-slot-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Scheduler component.
 *
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { SchedulerModule } from '@progress/kendo-angular-scheduler';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, SchedulerModule],
 *     bootstrap:    [AppComponent]
 * })
 *
 * export class AppModule {}
 * ```
 */
export class SchedulerModule {
}
SchedulerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: SchedulerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SchedulerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.10", ngImport: i0, type: SchedulerModule, imports: [i1.SchedulerComponent, i2.SchedulerCustomMessagesComponent, i3.SchedulerViewDirective, i4.DataBindingDirective, i5.SlotSelectableDirective, i6.PDFComponent, i7.PDFCommandDirective, i8.ShortcutsDirective, i9.AgendaViewComponent, i10.DayViewComponent, i11.MonthViewComponent, i12.MultiDayViewComponent, i13.MultiWeekViewComponent, i14.TimelineMonthViewComponent, i15.TimelineViewComponent, i16.TimelineWeekViewComponent, i17.WeekViewComponent, i18.WorkWeekViewComponent, i19.YearViewComponent, i20.ReactiveEditingDirective, i21.RecurrenceEditorComponent, i22.TimeZoneEditorComponent, i23.ToolbarNavigationComponent, i24.ToolbarTemplateDirective, i25.ToolbarViewSelectorComponent, i26.AgendaDateTemplateDirective, i27.AgendaTimeTemplateDirective, i28.AllDayEventTemplateDirective, i29.AllDaySlotTemplateDirective, i30.DateHeaderTemplateDirective, i31.EventTemplateDirective, i32.EditDialogTemplateDirective, i33.GroupHeaderTemplateDirective, i34.MajorTimeHeaderTemplateDirective, i35.MinorTimeHeaderTemplateDirective, i36.MonthDaySlotTemplateDirective, i37.MultiWeekDaySlotTemplateDirective, i38.TimeSlotTemplateDirective], exports: [i1.SchedulerComponent, i2.SchedulerCustomMessagesComponent, i3.SchedulerViewDirective, i4.DataBindingDirective, i5.SlotSelectableDirective, i6.PDFComponent, i7.PDFCommandDirective, i8.ShortcutsDirective, i9.AgendaViewComponent, i10.DayViewComponent, i11.MonthViewComponent, i12.MultiDayViewComponent, i13.MultiWeekViewComponent, i14.TimelineMonthViewComponent, i15.TimelineViewComponent, i16.TimelineWeekViewComponent, i17.WeekViewComponent, i18.WorkWeekViewComponent, i19.YearViewComponent, i20.ReactiveEditingDirective, i21.RecurrenceEditorComponent, i22.TimeZoneEditorComponent, i23.ToolbarNavigationComponent, i24.ToolbarTemplateDirective, i25.ToolbarViewSelectorComponent, i26.AgendaDateTemplateDirective, i27.AgendaTimeTemplateDirective, i28.AllDayEventTemplateDirective, i29.AllDaySlotTemplateDirective, i30.DateHeaderTemplateDirective, i31.EventTemplateDirective, i32.EditDialogTemplateDirective, i33.GroupHeaderTemplateDirective, i34.MajorTimeHeaderTemplateDirective, i35.MinorTimeHeaderTemplateDirective, i36.MonthDaySlotTemplateDirective, i37.MultiWeekDaySlotTemplateDirective, i38.TimeSlotTemplateDirective] });
SchedulerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: SchedulerModule, providers: [
        ToolbarService,
        ViewContextService,
        ViewStateService,
        EditService,
        IconsService,
        PopupService,
        ResizeBatchService,
        // DateInputsModule providers
        CalendarDOMService,
        CenturyViewService,
        DecadeViewService,
        MonthViewService,
        YearViewService,
        NavigationService,
        TimePickerDOMService,
        HoursService,
        MinutesService,
        SecondsService,
        MillisecondsService,
        DayPeriodService,
        // DropDownsModule providers
        DialogContainerService,
        DialogService,
        WindowService,
        WindowContainerService,
    ], imports: [i1.SchedulerComponent, i2.SchedulerCustomMessagesComponent, i6.PDFComponent, i7.PDFCommandDirective, i9.AgendaViewComponent, i10.DayViewComponent, i11.MonthViewComponent, i12.MultiDayViewComponent, i13.MultiWeekViewComponent, i14.TimelineMonthViewComponent, i15.TimelineViewComponent, i16.TimelineWeekViewComponent, i17.WeekViewComponent, i18.WorkWeekViewComponent, i19.YearViewComponent, i21.RecurrenceEditorComponent, i22.TimeZoneEditorComponent, i23.ToolbarNavigationComponent, i25.ToolbarViewSelectorComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.10", ngImport: i0, type: SchedulerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_SCHEDULER],
                    exports: [...KENDO_SCHEDULER],
                    providers: [
                        ToolbarService,
                        ViewContextService,
                        ViewStateService,
                        EditService,
                        IconsService,
                        PopupService,
                        ResizeBatchService,
                        // DateInputsModule providers
                        CalendarDOMService,
                        CenturyViewService,
                        DecadeViewService,
                        MonthViewService,
                        YearViewService,
                        NavigationService,
                        TimePickerDOMService,
                        HoursService,
                        MinutesService,
                        SecondsService,
                        MillisecondsService,
                        DayPeriodService,
                        // DropDownsModule providers
                        DialogContainerService,
                        DialogService,
                        WindowService,
                        WindowContainerService,
                    ],
                }]
        }] });
